ergo-nipopow/src/nipopow_proof.rs

```diff
@@ -0,0 +1,199 @@
+            m,
+            k,
+            prefix,
+            suffix_head,
+            suffix_tail,
+        }
+    }
+
+    /// Implementation of the ≥ algorithm from [`KMZ17`], see Algorithm 4
+    ///
+    /// [`KMZ17`]: https://fc20.ifca.ai/preproceedings/74.pdf
+    pub fn is_better_than(&self, that: &NipopowProof) -> bool {
+        if self.is_valid() && that.is_valid() {
+            if let Some(lca) = self.popow_algos.lowest_common_ancestor(
+                &self.headers_chain().collect::<Vec<_>>(),
+                &that.headers_chain().collect::<Vec<_>>(),
+            ) {
+                let self_headers = self
+                    .headers_chain()
+                    .filter(|h| h.height > lca.height)
+                    .collect::<Vec<_>>();
+                let that_headers = that
+                    .headers_chain()
+                    .filter(|h| h.height > lca.height)
+                    .collect::<Vec<_>>();
+                self.popow_algos.best_arg(&self_headers, self.m)
+                    > self.popow_algos.best_arg(&that_headers, self.m)
+            } else {
+                false
+            }
+        } else {
+            self.is_valid()
+        }
+    }
+
+    fn is_valid(&self) -> bool {
+        self.has_valid_connections() && self.has_valid_heights()
+    }
+
+    /// Checks the connections of the blocks in the proof. Adjacent blocks should be linked either
+    /// via interlink or parent block id. Returns true if all adjacent blocks are correctly
+    /// connected.
+    fn has_valid_connections(&self) -> bool {
+        self.prefix
+            .iter()
+            .zip(
+                self.prefix
+                    .iter()
+                    .skip(1)
+                    .chain(std::iter::once(&self.suffix_head)),
+            )
+            .all(|(prev, next)| next.interlinks.contains(&prev.header.id))
+            && std::iter::once(&self.suffix_head.header)
+                .chain(self.suffix_tail.iter())
+                .zip(self.suffix_tail.iter())
+                .all(|(prev, next)| next.parent_id == prev.id)
+    }
+
+    /// Checks if the heights of the header-chain provided are consistent, meaning that for any two
+    /// blocks b1 and b2, if b1 precedes b2 then b1's height should be smaller. Return true if the
+    /// heights of the header-chain are consistent
+    fn has_valid_heights(&self) -> bool {
+        self.headers_chain()
+            .zip(self.headers_chain().skip(1))
+            .all(|(prev, next)| prev.height < next.height)
+    }
+
+    /// Returns an iterator representing a chain of `Headers` from `self.prefix`, to
+    /// `self.suffix_head` and `self.suffix_tail`.
+    pub(crate) fn headers_chain(&self) -> impl Iterator<Item = &Header> {
+        self.prefix
+            .iter()
+            .map(|p| &p.header)
+            .chain(std::iter::once(&self.suffix_head.header).chain(self.suffix_tail.iter()))
+    }
+}
+
+impl ScorexSerializable for NipopowProof {
```

Comment:
Ideally, we should test the serialization roundtrip (with `Arbitrary` impl, etc.) but without `PoPowHeader` it's not possible. We don't have to wait probably and merge it as is with a new issue to test serialization. What do you think?

Reply:
I agree. I've created an issue for it and assigned myself to the task.